### PR TITLE
feat: Add option for reset/backup from different environment to Django admin

### PR DESCRIPTION
**Pull request by david-code**
## Description

This PR adds a reset button to the Django admin interface that can be optionally enabled. When pressed, it will trigger the following actions:

- S3 buckets from the source environment (A) will be synced with buckets in the target environment (B)
- A job will be triggered on A's Backend Service to create a backup archive and upload it to another S3 bucket.
- This archive will be downloaded by B's Backend Service. B's database will be wiped and restored from the archive. For all object in B's Database, URLs pointing to S3 buckets in A will be updated to point to the corresponding URLs in B.

There are a few tricky parts here.

- We use the database to save the progress of the long-running jobs, but the database needs to be wiped. This means that the background jobs table cannot be wiped.
- While we delete the database, the user will continue to query back to an endpoint. This means that Django's authentication will not work as we want it to if the user queries before the user object is re-added to the database.

### CODE REVIEW

**Comment by paulschreiber** [comment]
How do you want to handle cases where the hostname changes between instances? That will affect stored URLs.

**Comment by david-code** [comment]
> How do you want to handle cases where the hostname changes between instances? That will affect stored URLs.

Good point, I hadn't thought of that. I will take another look and see what the options could be.

File: terraso_backend/apps/core/management/commands/backup.py
Comment (paulschreiber) [inline terraso_backend/apps/core/management/commands/backup.py:82]: issue: please explain what you are cleaning up.

File: terraso_backend/apps/core/management/commands/backup.py
Comment (david-code) [reply terraso_backend/apps/core/management/commands/backup.py:37]: I think I originally wanted to write the migrations JSON directly to the backup JSON file. I realized later that this would be way more trouble than it's worth. I'll remove the comment.

File: terraso_backend/apps/core/management/commands/loadbackup.py
Comment (david-code) [reply terraso_backend/apps/core/management/commands/loadbackup.py:50]: I used the [django-storages](https://django-storages.readthedocs.io/en/latest/backends/amazon-S3.html) library here because I saw it used in other parts of the code. This implements the Django storage API, which doesn't provide an API to download files to the local file system. However, poking around a bit, there is a way to use the underlying `boto3` library, which should allow this. I will look into this more.

File: terraso_backend/apps/core/management/commands/loadbackup.py
Comment (david-code) [reply terraso_backend/apps/core/management/commands/loadbackup.py:42]: It's just to grab the migrations JSON and backup JSON file. I will make this more explicit (I think it is in the other method)

File: terraso_backend/apps/core/management/commands/loadbackup.py
Comment (paulschreiber) [inline terraso_backend/apps/core/management/commands/loadbackup.py:43]: question: what else is in the files array?

Comment (david-code) [reply terraso_backend/apps/core/management/commands/loadbackup.py:43]: It's possible that we have more than one set of backups in the bucket, for convenience or if there is a backup we want to save for whatever reason. Do you think there is a better way to handle this?

File: .env.sample
Comment (josebui) [inline .env.sample:33]: Suggestion: Can we have a `dev` bucket name named like the others, the have this structure: `files.dev.terraso.org`

**Comment by josebui** [comment]
Question: @david-code Can you add a comment with some example usages of the 2 commands? or maybe even better can you add those to the README?

**Comment by josebui** [comment]
Question: @david-code Can this hold multiple backups for one environment in the same S3 bucket?

**Comment by david-code** [comment]
> Question: @david-code Can you add a comment with some example usages of the 2 commands? or maybe even better can you add those to the README?

Yes, good idea. I will do that.

> Question: @david-code Can this hold multiple backups for one environment in the same S3 bucket?

Yes, but right now it will only use the most recent files. The files are named according to the date created and can be sorted alphabetically. I was thinking of adding an option to the backup script to allow deleting olders backups so we don't have to do that by hand.

**Comment by david-code** [comment]
> How do you want to handle cases where the hostname changes between instances? That will affect stored URLs.

I have added a new commit to address the issue of rewriting the URLs in the database. It is a bit complicated, and will require basically copying the S3 buckets to the restore instance buckets afterwards. I'll add some information in the README.

File: backup/Dockerfile
Comment (paulschreiber) [inline backup/Dockerfile:10]:
```suggestion
ENTRYPOINT ["/script/backup.sh"]

```

File: backup/url_rewrites.conf
Comment (paulschreiber) [inline backup/url_rewrites.conf:7]: issue: add newline to end of file

File: backup/env.sample
Comment (paulschreiber) [inline backup/env.sample:6]: issue: add newline to end of file

File: terraso_backend/config/settings.py
Comment (paulschreiber) [inline terraso_backend/config/settings.py:226]: issue: update to DB_BACKUP_S3_BUCKET

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: terraso_backend/custom_templates/admin/index.html
Comment (paulschreiber) [inline terraso_backend/custom_templates/admin/index.html:66]: Please put this in a standalone JS file so it can be linted/minified.

File: terraso_backend/custom_templates/admin/index.html
Comment (paulschreiber) [inline terraso_backend/custom_templates/admin/index.html:18]: question: do you need to add this manually?

File: terraso_backend/apps/core/views.py
Comment (paulschreiber) [inline terraso_backend/apps/core/views.py:80]: question: where does 15 seconds come from?
suggestion: make `SLEEP_TIME` a constant.

File: terraso_backend/apps/core/views.py
Comment (paulschreiber) [inline terraso_backend/apps/core/views.py:83]:
```suggestion
```

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: terraso_backend/apps/core/static/instance_restore.js
Comment (paulschreiber) [inline terraso_backend/apps/core/static/instance_restore.js:8]: Won't `if (document.cookie)` do the same thing?

File: terraso_backend/apps/core/static/instance_restore.js
Comment (paulschreiber) [inline terraso_backend/apps/core/static/instance_restore.js:56]: suggestion: 'Job started' (or other nontechnical language).

File: terraso_backend/apps/core/management/commands/loadbackup.py
Comment (david-code) [inline terraso_backend/apps/core/management/commands/loadbackup.py:253]: This code creates a new user with the same ID as the currently logged in user. This is just to make sure that the user ID of the user triggering the reset does not change and they are not logged out. We copy any other information that might have changed from the user in the DB being restored.

File: terraso_backend/apps/core/management/commands/loadbackup.py
Comment (david-code) [inline terraso_backend/apps/core/management/commands/loadbackup.py:211]: These migrations need to be skipped. `23` adds the background tasks table, but this table needs to not be deleted. `25` adds taxonomy terms that are already present in the data restore. In the future, we can add other migrations that need to be skipped here.

File: terraso_backend/apps/core/views.py
Comment (david-code) [inline terraso_backend/apps/core/views.py:90]: The user triggering the restore job will be deleted from the database at some point. This endpoint checks the cookie to see if the user is the person who triggered the job, and only returns the status if this is the case.

File: .env.sample
Comment (josebui) [inline .env.sample:39]: Issue: Missing the last line break

File: terraso_backend/apps/core/management/README.md
Comment (josebui) [inline terraso_backend/apps/core/management/README.md:24]: Question: Why is the DB_RESTORE_CONFIG_FILE value configurable? In what situations will this be changed?

File: Dockerfile
Comment (david-code) [reply Dockerfile:9]: Good point, I will check it out.

File: terraso_backend/apps/core/models/landscapes.py
Comment (david-code) [reply terraso_backend/apps/core/models/landscapes.py:46]: I think I rebased onto staging to make sure that I had the most recent changes for testing. This is a good point. It might make sense to merge this after the profile landscape changes are merged to main.

File: terraso_backend/apps/core/management/README.md
Comment (david-code) [reply terraso_backend/apps/core/management/README.md:24]: My reasoning was
- Don't want to hard code URLs into the code
- Don't want to add like 5 new env variables to the config file just for the backup, which will only be used on the testing environment

It is pretty easy to deploy it and is just a bit cleaner in my opinion.

**Comment by david-code** [comment]
I got rid of the earlier commits that I got when I rebase on the `staging` branch. However this means that the migration I introduced will probably cause problems when Jose's PR for the landscape profile is merged. @josebui just a heads up on this.

File: terraso_backend/custom_templates/admin/index.html
Comment (paulschreiber) [inline terraso_backend/custom_templates/admin/index.html:7]:
```suggestion
    <button id="restore-button" type="button">Load data</button>
```

File: terraso_backend/apps/core/management/commands/loadbackup.py
Comment (paulschreiber) [inline terraso_backend/apps/core/management/commands/loadbackup.py:248]: issue: please write a detailed comment explaining what is going on with the user ids, which tables are skipped, etc.

File: terraso_backend/apps/core/views.py
Comment (paulschreiber) [inline terraso_backend/apps/core/views.py:106]: suggestion: move this to the top of the file and give it a more descriptive name.

File: terraso_backend/apps/core/views.py
Comment (paulschreiber) [inline terraso_backend/apps/core/views.py:164]: suggestion: if you are going to leave this in production, add more details.

**Review by paulschreiber — CHANGES_REQUESTED** [review]

File: terraso_backend/apps/core/static/instance_restore.js
Comment (david-code) [reply terraso_backend/apps/core/static/instance_restore.js:8]: I think so, but I copied this function directly from the [Django docs](https://docs.djangoproject.com/en/4.1/howto/csrf/#acquiring-the-token-if-csrf-use-sessions-and-csrf-cookie-httponly-are-false) so I'm not sure if there's a corner case that the double check covers.

File: terraso_backend/custom_templates/admin/index.html
Comment (paulschreiber) [inline terraso_backend/custom_templates/admin/index.html:8]:
```suggestion
    <p>Replace with data from another Terraso instance.</p>
```

File: terraso_backend/apps/core/management/commands/loadbackup.py
Comment (paulschreiber) [inline terraso_backend/apps/core/management/commands/loadbackup.py:247]:
```suggestion
                    # triggering the reset) retains the same user ID in the restored database.
```

File: terraso_backend/apps/core/management/commands/loadbackup.py
Comment (paulschreiber) [inline terraso_backend/apps/core/management/commands/loadbackup.py:275]:
```suggestion
            if not config.is_file() :
                raise CommandError(
                    f"Path supplied for URL rewrites is not a file: {str(config)}. "
                    "URL rewrites not applied."
                )
            if not os.access(config, os.R_OK) :
                raise CommandError(
                    f"Cannot read config file: {str(config)}. "
                    "URL rewrites not applied."
                )

```

**Review by paulschreiber — CHANGES_REQUESTED** [review]

**Review by paulschreiber — APPROVED** [review]

**Pull request merged**
